Reader Editor: Focus Editor on FoldableCard open

DustyReagan opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/loop/issues/514

## Proposed Changes

* Introduce a `focusEditor` function.
* Call `focusEditor` when the `FoldableCard` component is opened.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We can assume the user's intent to write once they've clicked the "Write a new post" FoldableCard CTA. Giving the Editor focus makes the toolbar appear, and there is one less click to start writing.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* On Calypso Live go to /reader
* Click "Write a new post"
* You should see the Editor toolbar, and it should be focused.
* Ensure you can type the letters `j` and `k` in the Editor.
* Click outside of the Editor so it loses focus and press the `j` and `k` keys. They should act as keyboard shortcuts.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

DustyReagan committed:
Focus editor on FoldableCard open

DustyReagan commented on client/reader/components/quick-post/index.tsx:
```diff
@@ -105,7 +105,7 @@ function QuickPost( {
 						dispatch(
 							receiveNewPost( {
-								streamKey: `following`,
+								streamKey: 'following',
```
Just making the linter happy.

Addison-Stavlo commented on client/reader/components/quick-post/utils.tsx:
```diff
@@ -1,12 +1,48 @@
+ * Focus the Editor iframe if possible.
+ * @returns {void}
+ */
+export const focusEditor = (): void => {
```
is it worth checking `isEditorIframeFocused` first for an early return here?

DustyReagan replied:
I think so! Added.

Addison-Stavlo commented on client/reader/components/quick-post/utils.tsx:
```diff
@@ -1,12 +1,48 @@
+			return false;
+		}
+		editable.focus();
+		editable.dispatchEvent( new MouseEvent( 'click' ) );
```
why do we need to both focus it and dispatch a click?  will dispatching a click ruin a11y-focus styles that only show up when navigating with keyboard flows and disappear with mouse interaction? (edit- well we dont have a11y-focus styles around here yet, and are going into the editor paragraph anyways so that point may be moot).

DustyReagan replied:
It looks like `focus` is enough. I removed the `dispatchEvent`.

Addison-Stavlo reviewed (commented):
I have mixed thoughts and feelings here:
* The editor looks so empty without focusing in to activate the toolbar.
* Focus jump still feels a little funny in keyboard flows as we do still move from the expected position and jump over some items.
* We are also skipping focus over the site selector with this which might make it more overlooked (?maybe? maybe more by a11y users?)

One oddity is if I have content in the editor already (closed it, and reopening), the focus goes into the first paragraph block and at the beginning of the selection:

<img width="349" alt="Screenshot 2025-03-07 at 9 11 28 AM" src="https://github.com/user-attachments/assets/31313aff-9f9e-448f-944a-2af9149add07" />


<br/>
Im also noticing some regressions on this branch from production, although I am not certain how these changes would cause them. Maybe a rebase with trunk will help:

* open the editor, type in first paragraph.
* press enter. 
* type in a second paragraph. notice the j shortkey jumps out of the editor (i dont notice this in production)

DustyReagan committed:
Fix keyboard shortcut regression

DustyReagan committed:
Return early if isEditorIframeFocused

DustyReagan committed:
Remove unnecessary CSS

DustyReagan commented:
Thanks for the review @Addison-Stavlo! After @artemiomorales explored other approaches and from our conversation in our team meeting, we'll push forward with this approach and iterate from there.

I've rebased and addressed a few of your concerns:

> One oddity is if I have content in the editor already (closed it, and reopening), the focus goes into the first paragraph block and at the beginning of the selection:

We could probably manage to place the cursor at the end of the user content on focus, but I'm not sure that's better than placing it at the beginning. I feel like the solution to this is to remember where the cursor was and put it back. For v1 I propose we leave this functionality as is and revisit this.

> Im also noticing some regressions on this branch from production, although I am not certain how these changes would cause them. Maybe a rebase with trunk will help:
> 
>     * open the editor, type in first paragraph.
> 
>     * press enter.
> 
>     * type in a second paragraph. notice the j shortkey jumps out of the editor (i dont notice this in production)

This was indeed a regression and should be fixed by the commit https://github.com/Automattic/wp-calypso/pull/100969/commits/b9cd5a3dc044b4b17347b3ba6d4cd276cbb149f6

Addison-Stavlo reviewed (approved):
That makes sense to me dusty. this works well!
